### PR TITLE
yaml configuration for generating genesis.json, keys, etc by homi

### PR DESCRIPTION
**Pull request by kjeom**
## Proposed changes

- It is for yaml configuration to generate genesis.json, keys, etc..
- Note
  - You can use the legacy pattern to use the `homi` binary. (ex. `./homi setup local --baobab-test --cn-num 1`...)
  - The flag pattern like `--baobab-test` is top priority than yaml.
  - Usage after this PR
    - You can easily use the `homi` like `./homi --homi-yaml homi.yaml`
    - If you want modify some flags for repetitively test, there are two solutions.
      - Fix the flags in the yaml file
      - Pass the flag like `./homi --homi-yaml homi.yaml --cn-num 3`. The passed flags have more priority.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: cmd/conf/homi/homi.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+reward-staking-interval: 86400
+reward-proposer-interval: 3600
+reward-minimum-stake: "2000000"
+lower-bound-base-fee: 25000000000
+upper-bound-base-fee: 750000000000
+gas-target: 30000000
+block-gas-limit: 60000000
+base-fee-denominator: 20
+
+# Istanbul consensus engine options (default engine)
```
Comment (kjhman21) [inline cmd/conf/homi/homi.yaml:74]: Can we make this hierachical? like:
```
Istanbul:
  epoch:604800
  proposer-policy: 0
  subgroup: 21
```

Comment (kjeom) [reply cmd/conf/homi/homi.yaml:74]: Yes, but to do that, genesis.json generation logic should be updated [code](https://github.com/klaytn/klaytn/pull/1661/files#diff-593565b447ca4aa305be6b967d5878f61840cc041d4bda82d7fca95234df3779L584).
We have the clique engine, but we can't select the clique engine in `baobab-test`. The default engine in cypress, cypress-test, baobab, ... is istanbul.

File: cmd/conf/homi/homi.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+test-num: 0
+validators-num: 0
+
+######################################################################
+#                       For deploy environment                       #
+######################################################################
+
+# Only for docker gen-type
+docker-image-id: klaytn/klaytn:latest
+subgroup-size: 21
```
Comment (ian0371) [inline cmd/conf/homi/homi.yaml:28]: We can remove this line. `subGroupSizeFlag` is not used.

File: cmd/homi/setup/cmd.go [hunk 3]
```diff
@@ -530,10 +535,10 @@ func RandStringRunes(n int) string {
 	return string(b)
 }
 
-func gen(ctx *cli.Context) error {
+func Gen(ctx *cli.Context) error {
 	genType := findGenType(ctx)
+	fmt.Println(genType)
```
Comment (JayChoi1736) [inline cmd/homi/setup/cmd.go:540]: is this line intentional?

Comment (kjeom) [reply cmd/homi/setup/cmd.go]: Ohh, It was for debug... redundant code. Thanks

File: cmd/conf/homi/homi.yaml (hunk 2 above)
Comment (kjeom) [reply cmd/conf/homi/homi.yaml]: Thanks!

File: cmd/homi/setup/cmd.go [hunk 4]
```diff
@@ -1091,3 +1118,21 @@ func removeSpacesAndLines(b []byte) string {
 	out = strings.Replace(out, "\n", "", -1)
 	return out
 }
+
+func homiFlagsFromYaml(ctx *cli.Context) error {
+	filePath := ctx.String(homiYamlFlag.Name)
+	fmt.Println(filePath)
```
Comment (JayChoi1736) [inline cmd/homi/setup/cmd.go:1124]: i think this is also debug msg.

Comment (kjeom) [reply cmd/homi/setup/cmd.go]: Thanks!

**Review by JayChoi1736 — APPROVED** [review]

**Review by kjhman21 — APPROVED** [review]

**Review by ian0371 — APPROVED** [review]

**Review by mckim19 — DISMISSED** [review]

File: cmd/homi/setup/cmd.go [hunk 5]
```diff
@@ -1064,22 +1081,32 @@ func WriteFile(content []byte, parentFolder string, fileName string) {
...
+	return false
+}
+
 func findGenType(ctx *cli.Context) int {
+	// Klaytn-Node: genTypeFlag's default value is docker
+	genTypeFlag := ctx.String(genTypeFlag.Name)
+	if ctx.Args().Present() && existGenTypeArg(ctx.Args()[0]) {
+		genTypeFlag = ctx.Args()[0]
+	}
 	genType := TypeNotDefined
```
Comment (aidan-kwon) [inline cmd/homi/setup/cmd.go:1099]: Isn't this enough?
```suggestion
	genType := TypeNotDefined
	if ctx.Args().Present() {
		genTypeFlag = ctx.Args()[0]
	}
```

Comment (kjeom) [reply cmd/homi/setup/cmd.go:1099]: I thought the `first argument` can be the unrelated option, not `genTpyeFlag`.
Is there anything I have missed?

File: cmd/homi/setup/cmd.go [hunk 6]
```diff
@@ -546,11 +550,24 @@ func gen(ctx *cli.Context) error {
...
+	// Note-klaytn : the following code that seems unnecessary is for the priority to flags, not yaml
+	if !baobab && !baobabTest && !cypress && !cypressTest && !serviceChain && !serviceChainTest && !clique {
+		genesisType := ctx.String(genesisTypeFlag.Name)
+		baobab = genesisType == "baobab"
+		baobabTest = genesisType == "baobab-test"
+		cypress = genesisType == "cypress"
+		cypressTest = genesisType == "cypress-test"
+		serviceChain = genesisType == "servicechain"
+		serviceChainTest = genesisType == "servicechain-test"
+		clique = genesisType == "clique"
```
Comment (henry-will) [inline cmd/homi/setup/cmd.go:568]:
```suggestion
	switch genesisType := ctx.String(genesisTypeFlag.Name); genesisType {
	case "baobab":
		baobab = true
	case "baobab-test":
		baobabTest = true
	case "cypress":
		cypress = true	
	case "cypress-test":
		cypressTest = true	
	case "servicechain":
		servicechain = true	
	case "servicechain-test":
		serviceChainTest = true		
	case "clique":
		clique = true		
	default:
		fmt.Printf("Unknown genesis type is %s.\n", genesisType)
	}
```
I like this switch-case style a bit more. You can decide at your convenience.

Comment (kjeom) [reply cmd/homi/setup/cmd.go:568]: Looks better thanks!

**Review by aidan-kwon — APPROVED** [review]

**Review by JayChoi1736 — APPROVED** [review]

**Review by ian0371 — APPROVED** [review]

**Review by kjhman21 — APPROVED** [review]

File: cmd/homi/setup/cmd.go [hunk 7]
```diff
@@ -1065,22 +1092,25 @@ func WriteFile(content []byte, parentFolder string, fileName string) {
...
-			os.Exit(1)
-		}
 	} else {
 		genType = TypeDocker
 	}
+	if genType == TypeNotDefined {
+		fmt.Printf("Wrong Type : %s\nSupported Types : [docker, local, remote, deploy]\n\n", genTypeFlag)
+		cli.ShowSubcommandHelp(ctx)
+		os.Exit(1)
+	}
```
Comment (henry-will) [inline cmd/homi/setup/cmd.go:1113]:
```suggestion
func indexGenType(genTypeFlag string) int {
	for typeIndex, typeString := range Types {
		if genTypeFlag == typeString {
			return typeIndex
		}
	}
	return TypeNotDefined
}

func findGenType(ctx *cli.Context) int {
	var genType int
	if ctx.Args().Present() {
		genType = indexGenType(ctx.Args()[0])
	} else {
		genType = indexGenType(ctx.String(genTypeFlag.Name))
	}
	
	if genType == TypeNotDefined {
		fmt.Printf("Wrong Type : %s\nSupported Types : [docker, local, remote, deploy]\n\n", genTypeFlag)
		cli.ShowSubcommandHelp(ctx)
		os.Exit(1)
	}
	return genType
}
``` 
I think this case skips the line "genTypeFlag := ctx.String(genTypeFlag.Name)" and does not check genTypeFlag.  
So I suggest considering your initial intention that the first argument should have a higher precedence than genTypeFlag

Comment (kjeom) [reply cmd/homi/setup/cmd.go:1113]: Clean code :) Thanks!

**Review by henry-will — APPROVED** [review]

**Review by JayChoi1736 — APPROVED** [review]

**Pull request merged**